protocol/proto/encoder.go

```diff
@@ -22,63 +22,47 @@ package proto
 
 import (
 	"fmt"
-	"io"
 
 	"github.com/m3db/m3x/pool"
 )
 
 type encoder struct {
-	w              io.Writer
 	buffer         []byte
 	bytesPool      pool.BytesPool
 	maxMessageSize int
 }
 
 // NewEncoder creates a new encoder, the implementation is not thread safe.
-func NewEncoder(w io.Writer, opts BaseOptions) Encoder {
-	return newEncoder(w, opts)
-}
-
-func newEncoder(w io.Writer, opts BaseOptions) *encoder {
+func NewEncoder(opts BaseOptions) Encoder {
 	if opts == nil {
 		opts = NewBaseOptions()
 	}
 	pool := opts.BytesPool()
 	return &encoder{
-		w:              w,
 		buffer:         getByteSliceWithLength(sizeEncodingLength, pool),
 		bytesPool:      pool,
 		maxMessageSize: opts.MaxMessageSize(),
 	}
 }
 
-func (e *encoder) Encode(m Marshaler) error {
+func (e *encoder) Encode(m Marshaler) ([]byte, error) {
```

Comment:
This `Encode` API introduces coupling between the encoding logic and the way to access the encoded stream. I'd change the API to only return an error here and add a  `Bytes` or a `Stream` method that returns the currently encoded bytes. This way if you need to access the bytes you can just use the `Bytes` API without having to rely on `Encode`, which would be awkward to use for that purpose.

Also make sure to put a comment on the `Bytes` API to make it clear the returned bytes are only valid till the next `Encode` call and should be read only and not modified outside (wish there's a way to enforce this somehow..).

Reply:
Sounds good